### ISSUE TITLE
does not warn about unreachable code arising from always-true conditions

### ISSUE THREAD

**Issue by arctic-penguin** [microsoft/pyright#4005]
**Describe the bug**
Pyright does not warn if conditions in if-then-else are always true, especially if that makes other branches unreachable.

**To Reproduce**
```
def main():
    x: str = "abc"
    if len(x) >= 0:
        raise ValueError
    print("message")
```
No warning or error.

**Expected behavior**
Pyright warns that `print("message")` is unreachable. It should also warn if an any block in an `if-then-else` makes the subsequent branches unreachable.

**VS Code extension or command-line**
Reproducible as standalone and with `coc.nvim`. 

```
pyright 1.1.273
```

**Comment by erictraut** [comment]
This is expected behavior, not a bug. A type checker doesn't know that `len(x) >= 0` will always be true because it has no innate knowledge of string lengths or the internal behavior of the `len` function when it is passed a `str`.

Furthermore, unreachable code conditions in pyright are detected in pyright based on a code flow graph that is constructed prior to type analysis. (The code flow graph is needed as an input to the type analyzer.) If the type analyzer determines later that a condition is always true or always false, this does not affect the code flow graph, and unreachable code is not reported. Always-true and always-false conditions are taken into account only for expression forms that do not require type analysis — forms like `if True` or `if sys.version_info >= (3, 10):`.
